3rdParty/iresearch/core/formats/columnstore2.cpp

```diff
@@ -801,82 +910,177 @@ doc_iterator::ptr fixed_length_column::iterator(bool /*consolidation*/) const {
+  }
+
+  sparse_column(std::optional<std::string>&& name, bstring&& payload,
+                column_header&& hdr, column_index&& index,
+                const index_input& data_in,
+                compression::decompressor::ptr&& inflater,
+                encryption::stream* cipher, std::vector<column_block>&& blocks)
+      : column_base{std::move(name), std::move(payload),
+                    std::move(hdr),  std::move(index),
+                    data_in,         cipher},
+        blocks_{std::move(blocks)},
+        inflater_{std::move(inflater)} {
     assert(header().docs_count);
     assert(ColumnType::kSparse == header().type);
   }
 
   virtual doc_iterator::ptr iterator(bool /*consolidation*/) const override;
 
+  void make_buffered(
+      irs::index_input& in, const memory_accounting_f& memory_accounter,
+      range<std::unique_ptr<column_reader>> next_sorted_columns) override {
+    buffered_memory_accounter_ = memory_accounter;
+    auto& hdr = mutable_header();
+    if (!is_encrypted(hdr)) {
+      if (make_buffered_data<false>(hdr, in, blocks_, column_data_,
+                                    next_sorted_columns, nullptr)) {
+        buffered_input_ = std::make_unique<irs::bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()});
+      }
+    } else {
+      irs::remapped_bytes_ref_input::mapping mapping;
+      if (make_buffered_data<true>(hdr, in, blocks_, column_data_,
+                                   next_sorted_columns, &mapping)) {
+        buffered_input_ = std::make_unique<irs::remapped_bytes_ref_input>(
+            irs::bytes_ref{column_data_.data(), column_data_.size()},
+            std::move(mapping));
+      }
+    }
+    if (buffered_input_) {
+      reset_stream(buffered_input_.get());
+    }
+  }
+
  private:
+  static std::vector<column_block> read_blocks_sparse(const column_header& hdr,
+                                                      index_input& in);
+
   template<typename ValueReader>
   class payload_reader : private ValueReader {
    public:
     template<typename... Args>
-    payload_reader(
-        const column_block* blocks,
-        Args&&... args)
-      : ValueReader{ std::forward<Args>(args)... },
-        blocks_{blocks} {
-    }
+    payload_reader(const column_block* blocks, Args&&... args)
+        : ValueReader{std::forward<Args>(args)...}, blocks_{blocks} {}
 
     bytes_ref payload(doc_id_t i);
 
    private:
     const column_block* blocks_;
-  }; // payload_reader
+  };  // payload_reader
+
+  template<bool encrypted>
+  bool make_buffered_data(
+      column_header& hdr, irs::index_input& in,
+      std::vector<column_block>& blocks,
+      std::vector<irs::byte_type>& column_data,
+      range<std::unique_ptr<column_reader>> next_sorted_columns,
+      remapped_bytes_ref_input::mapping* mapping) {
+    // idx adr/block offset length source
+    std::vector <std::tuple<size_t, bool, size_t, size_t, size_t>> chunks;
```

Comment:
```suggestion
    std::vector <std::tuple<size_t, bool, size_t, size_t, size_t>> chunks;
```
Why don't use any size_t for bool flag?